File: runtime/src/main/java/com/guet/flexbox/widget/AsyncImageSpec.java
package com.guet.flexbox.widget;

import android.content.Context;
import android.text.TextUtils;

import com.facebook.litho.ComponentContext;
import com.facebook.litho.ComponentLayout;
import com.facebook.litho.Diff;
import com.facebook.litho.Output;
import com.facebook.litho.Size;
import com.facebook.litho.annotations.FromBoundsDefined;
import com.facebook.litho.annotations.MountSpec;
import com.facebook.litho.annotations.OnBoundsDefined;
import com.facebook.litho.annotations.OnCreateMountContent;
import com.facebook.litho.annotations.OnMeasure;
import com.facebook.litho.annotations.OnMount;
import com.facebook.litho.annotations.Prop;
import com.facebook.litho.annotations.PropDefault;
import com.facebook.litho.annotations.ShouldUpdate;
import com.facebook.litho.utils.MeasureUtils;

import static android.widget.ImageView.ScaleType;

@MountSpec(isPureRender = true, poolSize = 10)
final class AsyncImageSpec {

    @PropDefault
    static final ScaleType scaleType = ScaleType.FIT_XY;
    @PropDefault
    static final float imageAspectRatio = 1f;

    @OnCreateMountContent
    static AsyncMatrixDrawable onCreateMountContent(Context c) {
        return new AsyncMatrixDrawable(c);
    }

    @OnMeasure
    static void onMeasure(ComponentContext c,
                          ComponentLayout layout,
                          int widthSpec,
                          int heightSpec,
                          Size size,
                          @Prop(optional = true) float imageAspectRatio) {
        MeasureUtils.measureWithAspectRatio(
                widthSpec,
                heightSpec,
                imageAspectRatio,
                size
        );
    }

    @OnBoundsDefined
    static void onBoundsDefined(
            ComponentContext c,
            ComponentLayout layout,
            Output<Integer> layoutWidth,
            Output<Integer> layoutHeight,
            Output<Integer> horizontalPadding,
            Output<Integer> verticalPadding) {
        horizontalPadding.set(layout.getPaddingLeft() + layout.getPaddingRight());
        verticalPadding.set(layout.getPaddingTop() + layout.getPaddingBottom());
        layoutWidth.set(layout.getWidth());
        layoutHeight.set(layout.getHeight());
    }

    @OnMount
    static void onMount(ComponentContext c,
                        AsyncMatrixDrawable drawable,
                        @Prop CharSequence url,
                        @Prop(optional = true) ScaleType scaleType,
                        @Prop(optional = true) float radius,
                        @FromBoundsDefined int layoutWidth,
                        @FromBoundsDefined int layoutHeight,
                        @FromBoundsDefined int horizontalPadding,
                        @FromBoundsDefined int verticalPadding) {
        drawable.mount(
                url,
                layoutWidth,
                layoutHeight,
                horizontalPadding,
                verticalPadding,
                radius,
                scaleType
        );
    }

    @ShouldUpdate(onMount = true)
    static boolean shouldUpdate(
            @Prop(optional = true) Diff<ScaleType> scaleType,
            @Prop Diff<CharSequence> url) {
        return scaleType.getNext() != scaleType.getPrevious()
                || !TextUtils.equals(url.getPrevious(), url.getNext());
    }
}

File: core/src/main/java/com/guet/flexbox/widget/AsyncImageSpec.kt
package com.guet.flexbox.widget

import android.content.Context
import android.text.TextUtils
import android.widget.ImageView.ScaleType
import com.facebook.litho.*
import com.facebook.litho.annotations.*
import com.facebook.litho.utils.MeasureUtils

@MountSpec(isPureRender = true, poolSize = 10)
internal object AsyncImageSpec {

    @PropDefault
    val scaleType = ScaleType.FIT_XY
    @PropDefault
    val imageAspectRatio = 1f

    @OnCreateMountContent
    fun onCreateMountContent(c: Context): AsyncMatrixDrawable {
        return AsyncMatrixDrawable(c)
    }

    @OnMeasure
    fun onMeasure(c: ComponentContext,
                  layout: ComponentLayout,
                  widthSpec: Int,
                  heightSpec: Int,
                  size: Size,
                  @Prop(optional = true) imageAspectRatio: Float) {
        MeasureUtils.measureWithAspectRatio(
                widthSpec,
                heightSpec,
                imageAspectRatio,
                size
        )
    }

    @OnBoundsDefined
    fun onBoundsDefined(
            c: ComponentContext,
            layout: ComponentLayout,
            layoutWidth: Output<Int>,
            layoutHeight: Output<Int>,
            horizontalPadding: Output<Int>,
            verticalPadding: Output<Int>) {
        horizontalPadding.set(layout.paddingLeft + layout.paddingRight)
        verticalPadding.set(layout.paddingTop + layout.paddingBottom)
        layoutWidth.set(layout.width)
        layoutHeight.set(layout.height)
    }

    @OnMount
    fun onMount(c: ComponentContext,
                drawable: AsyncMatrixDrawable,
                @Prop url: CharSequence,
                @Prop(optional = true) scaleType: ScaleType,
                @Prop(optional = true) borderRadius: Float,
                @Prop(optional = true) borderColor: Int,
                @Prop(optional = true) borderWidth: Float,
                @FromBoundsDefined layoutWidth: Int,
                @FromBoundsDefined layoutHeight: Int,
                @FromBoundsDefined horizontalPadding: Int,
                @FromBoundsDefined verticalPadding: Int) {
        drawable.mount(
                url,
                layoutWidth,
                layoutHeight,
                horizontalPadding,
                verticalPadding,
                borderRadius,
                borderWidth,
                borderColor,
                scaleType
        )
    }

    @OnUnmount
    fun onUnmount(c: ComponentContext,
                  drawable: AsyncMatrixDrawable) {
        drawable.unmount()
    }

    @ShouldUpdate(onMount = true)
    fun shouldUpdate(
            @Prop(optional = true) scaleType: Diff<ScaleType>,
            @Prop url: Diff<CharSequence>,
            @Prop(optional = true) borderRadius: Diff<Float>,
            @Prop(optional = true) borderColor: Diff<Int>,
            @Prop(optional = true) borderWidth: Diff<Float>): Boolean {
        return (scaleType.previous != scaleType.next
                || !TextUtils.equals(url.previous, url.next)
                || borderRadius.previous != borderRadius.next
                || borderColor.previous != borderColor.next
                || borderWidth.previous != borderWidth.next)
    }
}
